Feat/chat 채팅 기능 개선 및 Redis / 인증 기능 추가

geonuu09 opened the pull request:
## 🔍 주요 변경 사항

1. **WebSocket 핸드셰이크 인터셉터 추가**:
   - JWT 토큰을 WebSocket 연결 시 검증하고, 사용자 ID를 WebSocket 세션에 저장하여 이후 메시지에서 사용자가 인증된 상태로 처리되도록 함.
   - WebSocket 메시지 처리 시 `Principal`을 사용하여 인증된 사용자 ID를 추출하고 메시지를 전송할 때 해당 사용자 정보 활용.

2. **Redis 캐시 적용**:
   - 채팅 메시지를 Redis에 캐시하고, 일정 시간이 지난 후 DB에 저장하는 배치 처리를 구현하여 성능을 최적화.
   - Redis 장애 시 DB로 바로 메시지를 저장할 수 있도록 fallback 메커니즘 추가.

3. **채팅방 예약 시간 체크 로직 적용**:
   - 채팅방 예약 시간과 종료 시간을 Redis를 통해 효율적으로 체크하도록 개선.
   - Redis 캐시를 통해 DB 조회를 최소화하여 성능 향상.


   

---

## 💡 변경 이유
- 채팅 메시지는 단시간 내에 많은 수의 요청이 발생할 수 있어, DB에 실시간 저장 시 병목 현상이 발생할 수 있습니다.
- 이를 방지하기 위해 메시지를 Redis에 임시 저장한 후, 일정 시간(5분) 간격으로 모아서 DB에 저장하는 배치 처리 방식을 적용하였습니다.
- 배치를 통해 성능을 향상시키고, 시스템 부하를 최소화할 수 있도록 개선했습니다.


---

## 🚀 개선 결과

- 추후 테스트 진행 예정.

geonuu09 committed:
refactor : 시스템 메시지 제거 및 중복 메소드 수정

geonuu09 committed:
feat: 채팅방 퇴장하기 API 추가

- 퇴장 시 멘토 또는 멘티의 채팅방 상대값 변경 -> 알림기능 활성화 예정

geonuu09 committed:
refactor : 멘토와 멘티를 ChatRoom 연관관계에 추가

geonuu09 committed:
refactor : 검증 로직 수정

geonuu09 committed:
fix : 불린 값 잘못 저장

geonuu09 committed:
refactor : 엔티티 재설정

geonuu09 committed:
refactor : 엔티티 재설정

geonuu09 committed:
feat: RedisConfig Template 추가

geonuu09 committed:
feat: 웹소켓 유저 인증 추가

geonuu09 committed:
remove : 사용하지 않는 클래스제거

geonuu09 committed:
feat : 웹소켓 채팅 입장 및 메시지 전송 로직 추가

- 기존의 작성된 로직에서 개선
- RedisCache를 적용하여 채팅룸 정보를 캐시하여 DB 접근 최소화
- RedisCache를 적용하여 채팅 메시지 전송시 캐싱 후 배치를 사용하여 DB로 저장
- RedisCache TTL 적용하여 무분별한 캐싱 방지
- Redis 장애 발생시 DB에 직접 저장

geonuu09 committed:
feat : 웹소켓 채팅 입장 및 메시지 전송 로직 추가

- 기존의 작성된 로직에서 개선
- RedisCache를 적용하여 채팅룸 정보를 캐시하여 DB 접근 최소화
- RedisCache를 적용하여 채팅 메시지 전송시 캐싱 후 배치를 사용하여 DB로 저장
- RedisCache TTL 적용하여 무분별한 캐싱 방지
- Redis 장애 발생시 DB에 직접 저장

geonuu09 committed:
feat : 웹소켓 채팅 입장 및 메시지 전송 로직 추가

- 기존의 작성된 로직에서 개선
- RedisCache를 적용하여 채팅룸 정보를 캐시하여 DB 접근 최소화
- RedisCache를 적용하여 채팅 메시지 전송시 캐싱 후 배치를 사용하여 DB로 저장
- RedisCache TTL 적용하여 무분별한 캐싱 방지
- Redis 장애 발생시 DB에 직접 저장

geonuu09 committed:
Merge branch 'develop' of https://github.com/BootTalk-Service/boot-talk-back into feat/chat

geonuu09 committed:
refactor : 유저 입장 시 1회만 입장 메시지 전송

shCHO9801 commented on boottalk/src/main/java/com/icandoit/boottalk/stomp_chat/repository/RedisChatMessageRepository.java:
```diff
@@ -0,0 +1,68 @@
+
+    private final RedisTemplate<String, Object> redisTemplate;
+
+    private static final String CHAT_MESSAGE_KEY_PREFIX = "chat:messages:";
```
추후에 레디스 키들 저장해놓는 자료형 하나 만들면 좋을것 같습니다

geonuu09 replied:
네! 그 방법도 좋을 것 같아요! 
공용 Redis key를 담는 걸까요? 아니면 제 파트에 key만을 저장해놓으면 될까요?

shCHO9801 replied:
공용 redis key를 상수로 다루는 파일 하나 있으면 추후에 좋을것 같다는 생각이 듭니다!

shCHO9801 commented on boottalk/src/main/java/com/icandoit/boottalk/stomp_chat/config/WebSocketHandshakeInterceptor.java:
```diff
@@ -0,0 +1,54 @@
+        if (bearerToken != null && bearerToken.startsWith("Bearer ")) {
+            return bearerToken.substring(7);
+        }
+        return null;
```
해당 부분이 조금더 구체적인 문제 발생이라면 에러 발생이나 문제가 될 수도 있지 않은지 궁금합니다!

geonuu09 replied:
52번 째 줄에 예외처리가 없어서 질문을 남겨놓으신 걸까요?? 
이 코드에서 에러가 발생될 수 있을 경우가 있는지 잘 모르겠습니다! 혹시 어떤 에러가 의심이 되실까요?!

shCHO9801 replied:
네네 맞습니다 토큰이 없는 경우에 어떠한 익셉션이 나타난다는걸 처리하는걸 제가 못찾았습니다!

geonuu09 replied:
네 위 유틸 메서드 호출부에 예외처리 코드 추가하였습니다!

shCHO9801 commented on boottalk/src/main/java/com/icandoit/boottalk/stomp_chat/controller/ChatRoomController.java:
```diff
@@ -49,11 +50,21 @@ public ResponseEntity<List<ChatRoomResponseDto>> getUserChatRooms(
 
+    // 채팅 조회
     @GetMapping("/{roomUuid}/messages")
     public ResponseEntity<List<MessageResponseDto>> getUserChatMessages(
```
페이징 처리가 필요하면 추가하면 좋을것 같습니다!

geonuu09 replied:
커피챗 한 세션이 30분으로 짧은 시간이기에 List로 받도록 해달라는 요청이 있어서 이렇게 구현하였습니다!

shCHO9801 replied:
아하 확인했습니다!

shCHO9801 reviewed (commented):
제 짧은 지식으로는 이정도 밖에 코멘트 드리지 못할것 같습니다!
고생하셨습니다!

geonuu09 committed:
feat: 채팅 타이핑 상태 전송 기능 추가

- 사용자가 채팅 입력 시 typing 상태를 서버에 전송하도록 구현
- WebSocket 기반의 /chat.typing 엔드포인트 처리
- SimpMessagingTemplate을 활용하여 상대 유저에게 타이핑 상태 브로드캐스트
- ChatTypingRequestDto 및 ChatTypingResponseDto 정의
- 실시간 타이핑 상태에 따라 프론트 UI에서 "입력 중" 표시 가능

jooyeon0907 commented on boottalk/src/main/java/com/icandoit/boottalk/stomp_chat/repository/RedisChatMessageRepository.java:
```diff
@@ -0,0 +1,68 @@
+        return cached.stream()
+            .filter(obj -> obj instanceof ChatMessage)
+            .map(obj -> (ChatMessage) obj)
+            .toList();
```
RedisTemplate<String, Object> 이
키가 chat:messages:* 이면 ChatMessage객체 저장
키가 chat:room:* 이면 ChatRoom 객체 저장하는데 

이처럼 키 패턴에 따라 저장 타입이 고정되어있으니,
RedisTemplate<String, ChatMessage> , RedisTemplate<String, ChatRoom>  로 역할 분리하면 형변환없이도 사용 가능하여 코드도 더 간결해질 것 같다는 의견 살포시 남깁니다...!

geonuu09 replied:
아 저는 이 RedisTemplate 이 <String, Object>로만 사용이 가능한 줄 알았어요...! 알려주셔서 감사합니다 👍🏻
바로 적용해보도록 하겠습니다!

jooyeon0907 commented on boottalk/src/main/java/com/icandoit/boottalk/stomp_chat/repository/RedisChatMessageRepository.java:
```diff
@@ -0,0 +1,68 @@
+
+        for (ChatMessage msg : messages) {
+            redisTemplate.opsForList().rightPush(key, msg);
+        }
```
제가 알기론.. 엔티티를 Redis에 직접 캐시하면,
lazy loading 된 값이 제대로 초기화되지 않아서 오류가 날 수 있고, JPA랑 관계 없는 객체가 관리되는 것처럼 보이면서 성능 문제가 생길 수 있다고  합니다. 그래서 보통 DTO 형태로 저장하는 것이 더 안전하다고 알고 있습니다.

그래도 ChatMessage 는 다른 엔티티와  관계를 맺지 않고, 단순 정보를 저장하는 용도라 Redis 에 @Entity 객체를 저장하는 것이 괜찮을 수도 있을 것 같네요🤔
(ChatRoom 같은 경우는 연관 관계가 있어서 Lazy Loading 이슈에 안전 할 수 있을지 싶네요..!)

위에서 말한대로 그런 이유로 엔티티를 저장하고 계신 걸 수도 있겠지만,
이 부분 참고용으로 알고 계시면 좋을 것 같아 말씀드려요😊

geonuu09 replied:
오, 이런 부분은 몰랐던 사실이에요! 알려주셔서 감사합니다 🙇‍♂️
생각해보니 Redis에 저장할 때는 엔티티와의 타입 차이도 고려해야 했는데, 제가 너무 깊이 생각하지 못하고 진행했던 것 같네요.
덕분에 DTO로 분리해서 관리하는 게 더 안전하다는 걸 배웠습니다!

jooyeon0907 reviewed (commented):
일정 메시지들을 Redis에 캐시 후, 주기적으로 DB 로 저장하는 방식이 시스템의 응답 속도도 빨라지고, 데이터 처리 과정도 원활하게 되며 효율적인 것 같네요👍
코드에 고생의 흔적이 많이 느껴지네요..수고하셨습니다!!

geonuu09 committed:
feat: redis Template Bean 추가

- RedisTemplate<String, ChatRoom>
-RedisTemplate<String, ChatMessage> template

geonuu09 committed:
refactor: redis template 수정에 따라 코드 간결화

geonuu09 committed:
refactor: 클래스 위치 수정 및 redis key 불일치 수정

- List<ChatMessage> messagesToFlush = redisChatRepository.getMessagesForBatch(roomUuid);

roomUuid 미 추가 -> 추가

geonuu09 committed:
refactor: Redis 캐시 메시지를 Entity → DTO 기반으로 리팩토링

- Redis 저장/조회 시 ChatMessage 대신 ChatMessageResponseDto 사용
- 배치 플러시 로직에서 DTO 변환 후 DB에 저장
- Redis 메시지 삭제 로직도 DTO 기준으로 변경

geonuu09 committed:
refactor: Redis 캐싱 시 ChatRoom 엔티티 대신 DTO 사용하도록 변경

- ChatRoom 엔티티를 ChatRoomResponseDto로 변환하여 Redis에 저장하도록 수정
- Redis에서 조회할 때도 DTO 형태로 받아서 처리
- 서비스 로직 내에서 Entity → DTO 변환 로직 추가

geonuu09 committed:
refactor : redis key를 한 곳에서 관리할 수 있도록 `RedisKeyPrefix` 클래스 생성 및 코드 리팩토링

geonuu09 committed:
refactor: 토큰 추출과정에서 null인 경우, 유효하지 않은 토큰인 경우 예외처리 추가
